boa/src/bytecompiler.rs

```diff
@@ -0,0 +1,674 @@
+        node::{GetConstField, GetField, Identifier, StatementList},
+        op::{AssignOp, BinOp, BitOp, CompOp, LogOp, NumOp, UnaryOp},
+        Const, Node,
+    },
+    value::{RcBigInt, RcString},
+    vm::{CodeBlock, Opcode},
+    Value,
+};
+
+use std::collections::HashMap;
+
+fn u32_to_array(value: u32) -> [u8; 4] {
+    // Safety: Transmuting a `u32` primitive to
+    // an array of 4 bytes is safe.
+    unsafe { std::mem::transmute(value) }
+}
+
+fn u16_to_array(value: u16) -> [u8; 2] {
+    // Safety: Transmuting a `u16` primitive to
+    // an array of 2 bytes is safe.
+    unsafe { std::mem::transmute(value) }
+}
+
+fn u64_to_array(value: u64) -> [u8; 8] {
+    // Safety: Transmuting a `u64` primitive to
+    // an array of 8 bytes is safe.
+    unsafe { std::mem::transmute(value) }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+enum Literal {
+    String(RcString),
+    BigInt(RcBigInt),
+}
+
+#[must_use]
+#[derive(Debug, Clone, Copy)]
+struct Label {
+    index: u32,
+}
+
+#[derive(Debug, Clone)]
+struct LoopControlInfo {
+    label: Option<Box<str>>,
+    loop_start: u32,
+    continues: Vec<Label>,
+    breaks: Vec<Label>,
+}
+
+#[derive(Debug, Clone, Copy)]
+enum Access<'a> {
+    Variable { name: &'a Identifier },
+    ByName { node: &'a GetConstField },
+    ByValue { node: &'a GetField },
+    This,
+}
+
+#[derive(Debug)]
+pub struct ByteCompiler {
+    code_block: CodeBlock,
+    literals_map: HashMap<Literal, u32>,
+    names_map: HashMap<RcString, u32>,
+    loops: Vec<LoopControlInfo>,
+}
+
+impl Default for ByteCompiler {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ByteCompiler {
+    /// Represents a placeholder address that will be patched later.
+    const DUMMY_ADDRESS: u32 = u32::MAX;
+
+    #[inline]
+    pub fn new() -> Self {
+        Self {
+            code_block: CodeBlock::new(),
+            literals_map: HashMap::new(),
+            names_map: HashMap::new(),
+            loops: Vec::new(),
+        }
+    }
+
+    #[inline]
+    fn get_or_insert_literal(&mut self, liternal: Literal) -> u32 {
+        if let Some(index) = self.literals_map.get(&liternal) {
+            return *index;
+        }
+
+        let value = match liternal.clone() {
+            Literal::String(value) => Value::from(value),
+            Literal::BigInt(value) => Value::from(value),
+        };
+
+        let index = self.code_block.literals.len() as u32;
+        self.code_block.literals.push(value);
+        self.literals_map.insert(liternal, index);
+        index
+    }
+
+    fn get_or_insert_name(&mut self, name: &str) -> u32 {
+        if let Some(index) = self.names_map.get(name) {
+            return *index;
+        }
+
+        let name: RcString = name.into();
+        let index = self.code_block.names.len() as u32;
+        self.code_block.names.push(name.clone());
+        self.names_map.insert(name, index);
+        index
+    }
+
+    fn next_opcode_location(&mut self) -> u32 {
```

Comment:
I would also `#[inline]` this, could bring some optimization on the bounds checking.